compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1377,21 +1377,8 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
         let ident = self.lower_ident(ident);
         let param_span = ident.span;
-        let span = bounds
-            .iter()
-            .fold(Some(param_span.shrink_to_hi()), |span: Option<Span>, bound| {
-                let bound_span = bound.span();
-                // We include bounds that come from a `#[derive(_)]` but point at the user's code,
-                // as we use this method to get a span appropriate for suggestions.
-                if !bound_span.can_be_used_for_suggestions() {
-                    None
-                } else if let Some(span) = span {
-                    Some(span.to(bound_span))
-                } else {
-                    Some(bound_span)
-                }
-            })
-            .unwrap_or(param_span.shrink_to_hi());
+        let span =
+            bounds.iter().fold(param_span.shrink_to_hi(), |span, bound| span.to(bound.span()));
```

Comment:
IMO, the issue is that we are trying to reconstruct a span during lowering, which will mix spans coming from macro different expansions.
Fortunately, we know what `SyntaxContext` produces the `:`, it's `GenericParam::colon_span`.
So there are 2 cases here:
- if `colon_span` is `None`, we should use `param_span.shrink_to_hi()`;
- if not, we should use it, and extend it with `bound.span().find_ancestor_in_same_ctxt(colon_span.ctxt())`.

Reply:
`bound.span().find_ancestor_in_same_ctxt(colon_span.ctxt())` goes horribly wrong when the colon also comes from a macro parameter:

https://github.com/rust-lang/rust/blob/d379a25e415e1291ced6c8bbf03c080d638ecc49/src/test/ui/rust-2018/edition-lint-infer-outlives-macro.fixed#L89-L96

I changed this to search only in the "parent span", which is either the entire generics list for (normal) generic args or the `impl Trait` span.